### PR TITLE
Chasm code cleanup

### LINKED ISSUE #29082: Having a Chasm collapse below you whilst Wings are activated still kills you.
Whilst having wings activated, if a chasm collapses below you, you still die.

### PR DESCRIPTION
**Pull request by ghost**
Chasms now have a list called `falling_atoms`. Any atom that begins falling into the chasm is added to that list, and any atom in this list is unable to begin falling again; this is to avoid duplicate falling animations, which can and do happen and may cause runtimes.

Additionally, all manual `FLYING` movement type checks have been replaced with `is_flying()`.

Fixes #29082.

### CODE REVIEW

File: code/game/turfs/simulated/chasm.dm [hunk 1]
```diff
@@ -10,6 +10,7 @@
 	icon_state = "smooth"
 	canSmoothWith = list(/turf/open/floor/fakepit, /turf/open/chasm)
 	density = TRUE //This will prevent hostile mobs from pathing into chasms, while the canpass override will still let it function like an open turf
+	var/list/falling_atoms //Atoms currently falling into the chasm
```
Comment (ChangelingRain) [inline code/game/turfs/simulated/chasm.dm]: should be /static so that all chasms share it

File: code/game/turfs/simulated/chasm.dm [hunk 2]
```diff
@@ -115,6 +119,7 @@
 			var/mob/living/L = AM
 			L.Knockdown(100)
 			L.adjustBruteLoss(30)
+	LAZYREMOVE(falling_atoms, AM)
```
Comment (ChangelingRain) [inline code/game/turfs/simulated/chasm.dm]: falling_atoms -= AM

File: code/game/turfs/simulated/chasm.dm [hunk 3]
```diff
@@ -79,6 +80,9 @@
 			INVOKE_ASYNC(src, .proc/drop, thing)
 
 /turf/open/chasm/proc/droppable(atom/movable/AM)
+	for(var/V in falling_atoms)
+		if(AM == V)
+			return 0
```
Comment (ChangelingRain) [inline code/game/turfs/simulated/chasm.dm]: then replace these three lines with `if(falling_atoms[AM]) return FALSE`

File: code/game/turfs/simulated/chasm.dm [hunk 4]
```diff
@@ -163,6 +168,8 @@
 		var/mob/living/silicon/robot/S = AM
 		qdel(S.mmi)
 
+	falling_atoms[AM] = FALSE
```
Comment (RemieRichards) [inline code/game/turfs/simulated/chasm.dm]: Congratulations! Now it won't garbage collect!

(```falling_atoms -= AM```)

File: code/game/turfs/simulated/chasm.dm [hunk 5]
```diff
@@ -115,6 +118,7 @@
 			var/mob/living/L = AM
 			L.Knockdown(100)
 			L.adjustBruteLoss(30)
+	falling_atoms[AM] = FALSE
```
Comment (RemieRichards) [inline code/game/turfs/simulated/chasm.dm]: ```falling_atoms -= AM```

Comment (ChangelingRain) [inline code/game/turfs/simulated/chasm.dm]: https://github.com/tgstation/tgstation/pull/29106#discussion_r126060517

File: code/game/turfs/simulated/chasm.dm [hunk 6]
```diff
@@ -79,33 +80,35 @@
 			INVOKE_ASYNC(src, .proc/drop, thing)
 
 /turf/open/chasm/proc/droppable(atom/movable/AM)
+	if(falling_atoms[AM])
+		return
```
Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: return FALSE

File: code/game/turfs/simulated/chasm.dm [hunk 7]
```diff
@@ -79,33 +80,35 @@
...
 	if(!isliving(AM) && !isobj(AM))
-		return 0
+		return
```
Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: return FALSE

File: code/game/turfs/simulated/chasm.dm [hunk 8]
```diff
@@ -79,33 +80,35 @@
...
 	if(istype(AM, /obj/singularity) || istype(AM, /obj/item/projectile) || AM.throwing)
-		return 0
+		return
```
Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: return FALSE

File: code/game/turfs/simulated/chasm.dm [hunk 9]
```diff
@@ -79,33 +80,35 @@
...
 	if(istype(AM, /obj/effect/portal))
 		//Portals aren't affected by gravity. Probably.
-		return 0
+		return
```
Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: return FALSE

File: code/game/turfs/simulated/chasm.dm [hunk 10]
```diff
@@ -79,33 +80,35 @@
...
 	//Flies right over the chasm
 	if(isliving(AM))
-		var/mob/MM = AM
-		if(MM.movement_type & FLYING)
-			return 0
+		var/mob/M = AM
+		if(M.is_flying())
+			return
```
Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: return FALSE

File: code/game/turfs/simulated/chasm.dm [hunk 11]
```diff
@@ -79,33 +80,35 @@
...
 	if(ishuman(AM))
 		var/mob/living/carbon/human/H = AM
 		if(istype(H.belt, /obj/item/device/wormhole_jaunter))
 			var/obj/item/device/wormhole_jaunter/J = H.belt
 			//To freak out any bystanders
 			visible_message("<span class='boldwarning'>[H] falls into [src]!</span>")
 			J.chasm_react(H)
-			return 0
-	return 1
+			return
```
Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: return FALSE

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/turfs/simulated/chasm.dm (hunk 6 above)
Comment (Xhuis) [inline code/game/turfs/simulated/chasm.dm]: Isn't that the same as just `return`?

Comment (Cyberboss) [inline code/game/turfs/simulated/chasm.dm]: yes, but readability

**Review by Cyberboss — APPROVED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**